### PR TITLE
feat(clerk-expo)!: React Native >= 0.73 and Expo Web support

### PR DESCRIPTION
**Pull request by octoper**
## Description

This PR introduces a new major release of the `@clerk/clerk-expo` SDK.

## Breaking changes

This release will require to have at least `react-native >= 0.73`, we are trying to be more consistent on which versions of React Native and Expo we support from now on.

## Features

This release will also add a couple features:

- Web support
- Prebuilt components for Web exported from `/web` subpath

Exported Components from `/web` subpath:
- `<SignIn/>`
- `<SignUp/>`
- `<UserProfile/>`
- `<OrganizationSwitcher/>`
- `<OrganizationProfile/>`
- `<OrganizationList/>`
- `<CreateOrganization/>`
- `<SignInButton/>`
- `<SignUpButton/>`
- `<SignOutButton/>`
- `<SignInWithMetamaskButton/>`
- `<GoogleOneTap/>` 

## Internal

The structure of `@clerk/clerk-expo` SDK has changes in order for us to support Native and Web platforms

 

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

**Comment by octoper** [comment]
!snapshot

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/clerk-expo | 1.2.0-snapshot.v0f4c367 |
| gatsby-plugin-clerk | 5.0.0-beta.45 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.2.0-snapshot.v0f4c367 --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/clerk-expo | 1.2.0-snapshot.v25f6d6e |
| gatsby-plugin-clerk | 5.0.0-beta.45 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.2.0-snapshot.v25f6d6e --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 0.0.2-snapshot.va3c6a5f |
| @clerk/clerk-expo | 1.3.0-snapshot.va3c6a5f |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/testing | 1.1.10-snapshot.va3c6a5f |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@0.0.2-snapshot.va3c6a5f --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.3.0-snapshot.va3c6a5f --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.1.10-snapshot.va3c6a5f --save-exact
```

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 0.0.2-snapshot.vda60bc7 |
| @clerk/clerk-expo | 1.3.0-snapshot.vda60bc7 |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/testing | 1.1.10-snapshot.vda60bc7 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@0.0.2-snapshot.vda60bc7 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.3.0-snapshot.vda60bc7 --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.1.10-snapshot.vda60bc7 --save-exact
```

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 0.0.2-snapshot.v0b9be88 |
| @clerk/clerk-expo | 1.3.0-snapshot.v0b9be88 |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/testing | 1.1.10-snapshot.v0b9be88 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@0.0.2-snapshot.v0b9be88 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.3.0-snapshot.v0b9be88 --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.1.10-snapshot.v0b9be88 --save-exact
```

**Comment by octoper** [comment]
!snapshot

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 0.0.3-snapshot.v738965a |
| @clerk/clerk-expo | 1.3.0-snapshot.v738965a |
| gatsby-plugin-clerk | 5.0.0-beta.45 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@0.0.3-snapshot.v738965a --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.3.0-snapshot.v738965a --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 0.0.3-snapshot.vcf5b726 |
| @clerk/clerk-expo | 1.3.0-snapshot.vcf5b726 |
| gatsby-plugin-clerk | 5.0.0-beta.45 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@0.0.3-snapshot.vcf5b726 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.3.0-snapshot.vcf5b726 --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 0.0.3-snapshot.ve796cf8 |
| @clerk/backend | 1.3.2-snapshot.ve796cf8 |
| @clerk/chrome-extension | 1.1.4-snapshot.ve796cf8 |
| @clerk/clerk-js | 5.9.0-snapshot.ve796cf8 |
| @clerk/elements | 0.10.3-snapshot.ve796cf8 |
| @clerk/clerk-expo | 1.3.0-snapshot.ve796cf8 |
| @clerk/express | 0.0.17-snapshot.ve796cf8 |
| @clerk/fastify | 1.0.19-snapshot.ve796cf8 |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/localizations | 2.4.8-snapshot.ve796cf8 |
| @clerk/nextjs | 5.2.4-snapshot.ve796cf8 |
| @clerk/clerk-react | 5.2.8-snapshot.ve796cf8 |
| @clerk/remix | 4.2.3-snapshot.ve796cf8 |
| @clerk/clerk-sdk-node | 5.0.16-snapshot.ve796cf8 |
| @clerk/shared | 2.3.3-snapshot.ve796cf8 |
| @clerk/tanstack-start | 0.1.7-snapshot.ve796cf8 |
| @clerk/testing | 1.1.11-snapshot.ve796cf8 |
| @clerk/themes | 2.1.12-snapshot.ve796cf8 |
| @clerk/types | 4.8.0-snapshot.ve796cf8 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@0.0.3-snapshot.ve796cf8 --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.3.2-snapshot.ve796cf8 --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.1.4-snapshot.ve796cf8 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.9.0-snapshot.ve796cf8 --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.10.3-snapshot.ve796cf8 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@1.3.0-snapshot.ve796cf8 --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.0.17-snapshot.ve796cf8 --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.19-snapshot.ve796cf8 --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@2.4.8-snapshot.ve796cf8 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.2.4-snapshot.ve796cf8 --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.2.8-snapshot.ve796cf8 --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.3-snapshot.ve796cf8 --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.16-snapshot.ve796cf8 --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.3.3-snapshot.ve796cf8 --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.1.7-snapshot.ve796cf8 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.1.11-snapshot.ve796cf8 --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.12-snapshot.ve796cf8 --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.8.0-snapshot.ve796cf8 --save-exact
```

File: integration/templates/expo-web/assets/fonts/SpaceMono-Regular.ttf
Comment (LekoArts) [inline integration/templates/expo-web/assets/fonts/SpaceMono-Regular.ttf:1]: Can we remove any non-essential assets from being committed to the repo? 🙏 Users who'll clone the repo in the future will thank us.

We can replace the font with the system font stack and the icons we can replace with e.g. an emoji or something

File: packages/expo/package.json [hunk 1]
```diff
@@ -44,24 +57,26 @@
...
     "tslib": "2.4.1"
   },
   "devDependencies": {
     "@clerk/eslint-config-custom": "*",
     "@types/base-64": "^1.0.2",
     "@types/node": "^20.11.24",
-    "@types/react": "*",
-    "@types/react-dom": "*",
+    "@types/react": "^18.2.0",
+    "@types/react-dom": "^18.2.0",
```
Comment (LekoArts) [inline packages/expo/package.json:68]: Can we keep those as `*`?

File: packages/expo/package.json [hunk 2]
```diff
@@ -44,24 +57,26 @@
...
+    "react-native": "^0.74.2",
     "typescript": "*"
   },
   "peerDependencies": {
     "expo-auth-session": ">=4",
     "expo-web-browser": ">=12.5.0",
     "react": ">=18",
-    "react-dom": ">=18"
+    "react-dom": ">=18",
+    "react-native": ">=0.73"
```
Comment (LekoArts) [inline packages/expo/package.json:79]: Can this new requirement break existing user apps that already use `@clerk/clerk-expo` and e.g. also have no intention of using Expo Web?

File: packages/expo/src/provider/singleton/createClerkInstance.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * // Somewhere in your code, outside of React you can do
+ * const token = await clerkInstance.session?.getToken();
+ * fetch('http://example.com/', {headers: {Authorization: token })
+ * @throws MissingPublishableKeyError publishableKey is missing and Clerk has not been initialized yet
+ * @returns HeadlessBrowserClerk
+ */
+export function createClerkInstance(ClerkClass: typeof Clerk) {
+  return (options?: BuildClerkOptions): HeadlessBrowserClerk | BrowserClerk => {
+    const { publishableKey = process.env.CLERK_PUBLISHABLE_KEY || '', tokenCache = MemoryTokenCache } = options || {};
```
Comment (LekoArts) [inline packages/expo/src/provider/singleton/createClerkInstance.ts:42]: Does this need the `EXPO_PUBLIC` prefix?

File: packages/expo/tsup.config.ts [hunk 4]
```diff
@@ -18,6 +18,7 @@ export default defineConfig(overrideOptions => {
     minify: false,
     sourcemap: true,
     legacyOutput: true,
+    skipNodeModulesBundle: true,
```
Comment (LekoArts) [inline packages/expo/tsup.config.ts:21]: Why do we need to add this?

**Review by LekoArts — CHANGES_REQUESTED** [review]

File: packages/expo/package.json (hunk 2 above)
Comment (octoper) [reply packages/expo/package.json:79]: Those have been added as we now need the `react-native` dep and the `react` one are added to override the momorepo's react version which does not match the requirement of the `react-native` dep. We had a problem with this as Renovate updates the momorepos React version but does not take account the Version needed for React Native

File: packages/expo/src/provider/singleton/createClerkInstance.ts (hunk 3 above)
Comment (octoper) [reply packages/expo/src/provider/singleton/createClerkInstance.ts:42]: Nice catch, probably didn't include it when I was doing rebase 😅

File: integration/templates/expo-web/assets/fonts/SpaceMono-Regular.ttf
Comment (octoper) [reply integration/templates/expo-web/assets/fonts/SpaceMono-Regular.ttf:1]: Makes sense, I will remove it!

File: packages/expo/package.json (hunk 2 above)
Comment (LekoArts) [reply packages/expo/package.json:79]: Yeah, I get that 👍 
What I tried to ask is:

Assume we have a user that currently has the latest `@clerk/clerk-expo` installed. Their app uses `react-native@0.72`. They now install this new version (from this PR) and suddenly their app breaks (either during installation because of npm peerDep mismatch or during runtime because `Platform` is not accessible).

Comment (octoper) [reply packages/expo/package.json:79]: Hmm you are right, let me check if we can lower that to `react-native@0.72` in order to avoid that, I changed it at first as it's a minor version change and I think that  nothing will break, but you are right that it might be caught on peerDep mismatches

**Comment by octoper** [comment]
!snapshot

File: packages/expo/tsup.config.ts (hunk 4 above)
Comment (octoper) [reply packages/expo/tsup.config.ts]: Nice catch, this was a leftover from a test I was doing, removed!

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.0.1-snapshot.v9559e57 |
| @clerk/backend | 1.4.0-snapshot.v9559e57 |
| @clerk/chrome-extension | 1.1.5-snapshot.v9559e57 |
| @clerk/clerk-js | 5.10.0-snapshot.v9559e57 |
| @clerk/elements | 0.10.5-snapshot.v9559e57 |
| @clerk/clerk-expo | 2.0.0-snapshot.v9559e57 |
| @clerk/express | 0.0.18-snapshot.v9559e57 |
| @clerk/fastify | 1.0.20-snapshot.v9559e57 |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/nextjs | 5.2.5-snapshot.v9559e57 |
| @clerk/clerk-react | 5.2.9-snapshot.v9559e57 |
| @clerk/remix | 4.2.4-snapshot.v9559e57 |
| @clerk/clerk-sdk-node | 5.0.17-snapshot.v9559e57 |
| @clerk/shared | 2.4.0-snapshot.v9559e57 |
| @clerk/tanstack-start | 0.1.8-snapshot.v9559e57 |
| @clerk/testing | 1.1.12-snapshot.v9559e57 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.0.1-snapshot.v9559e57 --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.4.0-snapshot.v9559e57 --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.1.5-snapshot.v9559e57 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.10.0-snapshot.v9559e57 --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.10.5-snapshot.v9559e57 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.0.0-snapshot.v9559e57 --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.0.18-snapshot.v9559e57 --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.20-snapshot.v9559e57 --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.2.5-snapshot.v9559e57 --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.2.9-snapshot.v9559e57 --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.4-snapshot.v9559e57 --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.17-snapshot.v9559e57 --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.4.0-snapshot.v9559e57 --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.1.8-snapshot.v9559e57 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.1.12-snapshot.v9559e57 --save-exact
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -136,7 +136,7 @@ jobs:
 
     strategy:
       matrix:
-        test-name: ['generic', 'express', 'quickstart', 'ap-flows', 'elements', 'sessions']
+        test-name: ['generic', 'express', 'quickstart', 'ap-flows', 'elements', 'sessions', 'expoWeb']
```
Comment (panteliselef) [inline .github/workflows/ci.yml:139]: nit: just to align it with `ap-flows`
```suggestion
        test-name: ['generic', 'express', 'quickstart', 'ap-flows', 'elements', 'sessions', 'expo-web']
```

File: .changeset/two-wombats-confess.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+```tsx
+import { SignUp } from "@clerk/clerk-expo/web";
+
+export default function Page() {
+  return <SignUp />;
+}
+```
+
+**Breaking change:** You need to use Expo 50 or later. The minimum required React Native version was bumped to `0.73`.
```
Comment (panteliselef) [inline .changeset/two-wombats-confess.md:17]: Should we mention the reason why we needed to bump ? Either here or in the docs ?

File: integration/testUtils/commonPageObject.ts [hunk 7]
```diff
@@ -5,8 +5,11 @@ export const common = ({ page }: TestArgs) => {
     continue: () => {
       return page.getByRole('button', { name: 'Continue', exact: true }).click();
     },
+    setEmailAddress: (val: string) => {
+      return self.getEmailAddressInput().fill(val);
+    },
     setPassword: (val: string) => {
-      return page.locator('input[name=password]').fill(val);
+      return self.getPasswordInput().fill(val);
```
Comment (panteliselef) [inline integration/testUtils/commonPageObject.ts:12]: ✅

File: package.json [hunk 8]
```diff
@@ -39,6 +39,7 @@
     "test:integration:generic": "E2E_APP_ID=react.vite.*,next.appRouter.withEmailCodes npm run test:integration:base -- --grep @generic",
     "test:integration:nextjs": "E2E_APP_ID=next.appRouter.* npm run test:integration:base -- --grep @nextjs",
     "test:integration:astro": "E2E_APP_ID=astro.* npm run test:integration:base -- --grep @astro",
+    "test:integration:expoWeb": "E2E_APP_ID=expo.expoWeb npm run test:integration:base -- --grep @expoWeb",
```
Comment (panteliselef) [inline package.json:42]: nit: same reason as above
```suggestion
    "test:integration:expo-web": "E2E_APP_ID=expo.expo-web npm run test:integration:base -- --grep @expo-web",
```

File: packages/expo/package.json [hunk 9]
```diff
@@ -44,7 +57,7 @@
     "@clerk/shared": "2.3.3",
     "@clerk/types": "4.8.0",
     "base-64": "^1.0.0",
-    "react-native-url-polyfill": "2.0.0",
+    "react-native-url-polyfill": "^2.0.0",
```
Comment (panteliselef) [inline packages/expo/package.json:60]: Usually we use specific versions, what's the reason behind this change ?

File: packages/expo/src/hooks/useOAuth.ts [hunk 10]
```diff
@@ -25,7 +27,7 @@ export type StartOAuthFlowReturnType = {
 export function useOAuth(useOAuthParams: UseOAuthFlowParams) {
   const { strategy } = useOAuthParams || {};
   if (!strategy) {
-    throw new Error('Missing oauth strategy');
+    throw errorThrower.throw('Missing oauth strategy');
```
Comment (panteliselef) [inline packages/expo/src/hooks/useOAuth.ts:30]: shouldn't this be
```suggestion
   errorThrower.throw('Missing oauth strategy');
```

File: packages/expo/tsconfig.json [hunk 11]
```diff
@@ -20,7 +20,8 @@
     "allowJs": true,
     "target": "ES2019",
     "noEmitOnError": false,
-    "incremental": true
+    "incremental": true,
+    "moduleSuffixes": [".ios", ".android", ".native", ""]
```
Comment (panteliselef) [inline packages/expo/tsconfig.json:24]: ✅

File: packages/expo/src/provider/singleton/singleton.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import { Clerk } from '@clerk/clerk-js/headless';
+
+import { createClerkInstance } from './createClerkInstance';
+
+/**
+ * @deprecated Use `getClerkInstance` instead. `Clerk` will be removed in the next major version.
+ */
+export { clerk } from './createClerkInstance';
```
Comment (panteliselef) [inline packages/expo/src/provider/singleton/singleton.ts:8]: I'd drop this, if we are cutting a new major

File: packages/expo/src/provider/singleton/singleton.web.ts [hunk 13]
```diff
@@ -0,0 +1,5 @@
+/**
+ * @deprecated Use `getClerkInstance` instead. `Clerk` will be removed in the next major version.
+ */
+export const clerk = globalThis?.window?.Clerk;
+export const getClerkInstance = globalThis?.window?.Clerk;
```
Comment (panteliselef) [inline packages/expo/src/provider/singleton/singleton.web.ts:5]: Do we need to export those in order to have parity between web and native ? 
❌ Also `getClerkInstance` for native is a function.

File: packages/expo/src/provider/ClerkProvider.tsx [hunk 14]
```diff
@@ -1,30 +1,40 @@
...
       // Force reset the state when the provided key changes, this ensures that the provider does not retain stale state.
       // See JS-598 for additional context.
       key={pk}
       {...rest}
       publishableKey={pk}
-      Clerk={getClerkInstance({ publishableKey: pk, tokenCache })}
-      standardBrowser={!isReactNative()}
+      sdkMetadata={SDK_METADATA}
+      Clerk={isNative() ? getClerkInstance({ publishableKey: pk, tokenCache }) : null}
+      standardBrowser={!isNative()}
```
Comment (panteliselef) [inline packages/expo/src/provider/ClerkProvider.tsx:37]: How does bundling work here ? 
`getClerkInstance` requires the `clerk-js` esm bundle, where as we actually only want to hotload on web. Does this mean we are loading clerk-js bundled and clerk-js hotload while on web ?

Or will metro drop `getClerkInstance` along side the import from clerk-js resulting in clerk-js only beeing hotloaded ?

**Review by panteliselef — COMMENTED** [review]
Very nice work, I've added a few comments, only one seems critical.

File: .changeset/two-wombats-confess.md (hunk 6 above)
Comment (octoper) [reply .changeset/two-wombats-confess.md:17]: This will be mentioned in the docs as part of release cycles for the SDK

File: packages/expo/src/provider/singleton/singleton.web.ts (hunk 13 above)
Comment (octoper) [reply packages/expo/src/provider/singleton/singleton.web.ts:5]: Yeah, we need them for parity so we don't have any differences between platforms.

Yeah you are right about `getClerkInstance` It's not used so that we can drop it

File: packages/expo/src/provider/ClerkProvider.tsx (hunk 14 above)
Comment (octoper) [reply packages/expo/src/provider/ClerkProvider.tsx:37]: Metro will handle the bundling here it will leverage platform-specific extensions and will either load `singleton` which bundles `clerk-js` or `singleton.web` which does not bundle `clerk-js` and makes the hot loading available.

File: packages/expo/src/provider/singleton/singleton.ts (hunk 12 above)
Comment (octoper) [reply packages/expo/src/provider/singleton/singleton.ts:8]: As we discussed we will not drop that in this version as we don't want to introduce any API changes, the bump of the major is only for React Native and Expo version bumps, to be more clear which version the SDK supports

File: packages/expo/package.json (hunk 9 above)
Comment (octoper) [reply packages/expo/package.json:60]: No reason for this change, nice catch 💯

File: packages/expo/src/provider/singleton/singleton.web.ts (hunk 13 above)
Comment (panteliselef) [reply packages/expo/src/provider/singleton/singleton.web.ts:5]: Not sure which will be dropping, can you confirm ?

Comment (octoper) [reply packages/expo/src/provider/singleton/singleton.web.ts:5]: We could drop it from the `singleton.web.tsx`, but I took a look again and it would be nicer to keep it just for having an understanding of how it works and what `getClerkInstance` does

File: packages/expo/package.json [hunk 15]
```diff
@@ -23,11 +23,24 @@
   },
   "license": "MIT",
   "author": "Clerk",
-  "main": "dist/index.js",
-  "source": "src/index.js",
-  "typings": "dist/index.d.ts",
+  "exports": {
+    ".": {
+      "types": "./dist/index.d.ts",
+      "import": "./dist/index.js",
```
Comment (LekoArts) [inline packages/expo/package.json:29]: We're only outputting CJS so you should combine `import` and `require` into `default`.

If we then ever publish CJS + ESM, we can split it into both again

File: packages/expo/src/provider/singleton/singleton.web.ts (hunk 13 above)
Comment (LekoArts) [reply packages/expo/src/provider/singleton/singleton.web.ts:5]: We should clear this up ☝️

File: packages/expo/src/web/uiComponents.tsx [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+  SignUpButton as BaseSignUpButton,
+  UserButton as BaseUserButton,
+  UserProfile as BaseUserProfile,
+} from '@clerk/clerk-react';
+import { Platform } from 'react-native';
+
+import { errorThrower } from '../utils/errors';
+
+const ErrorComponent = (componentName: string) => {
+  throw errorThrower.throw(`${componentName} component is not supported in the native environments.`);
```
Comment (LekoArts) [inline packages/expo/src/web/uiComponents.tsx:21]:
```suggestion
  errorThrower.throw(`${componentName} component is not supported in the native environments.`);
```

File: packages/expo/package.json (hunk 15 above)
Comment (octoper) [reply packages/expo/package.json]: Fixed!

File: packages/expo/package.json [hunk 17]
```diff
@@ -23,11 +23,24 @@
...
+      "require": "./dist/index.js"
+    },
+    "./web": {
+      "types": "./dist/web/index.d.ts",
+      "import": "./dist/web/index.js",
```
[the remark this answers is not in the document]
Comment (octoper) [reply packages/expo/package.json]: Fixed!

File: packages/expo/src/provider/singleton/singleton.web.ts (hunk 13 above)
Comment (octoper) [reply packages/expo/src/provider/singleton/singleton.web.ts:5]: Hmm, okay just did some tests.

We will need to keep this one as we export both `getClerkInstance` and `clerk`, and cause we want to keep it isomorphic if we remove this and either of these is used it will result in an error

**Review by LekoArts — APPROVED** [review]

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.0.3-snapshot.vc45c61c |
| @clerk/backend | 1.4.2-snapshot.vc45c61c |
| @clerk/elements | 0.10.6-snapshot.vc45c61c |
| @clerk/clerk-expo | 2.0.0-snapshot.vc45c61c |
| @clerk/express | 0.0.20-snapshot.vc45c61c |
| @clerk/fastify | 1.0.22-snapshot.vc45c61c |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/nextjs | 5.2.7-snapshot.vc45c61c |
| @clerk/remix | 4.2.6-snapshot.vc45c61c |
| @clerk/clerk-sdk-node | 5.0.19-snapshot.vc45c61c |
| @clerk/tanstack-start | 0.1.10-snapshot.vc45c61c |
| @clerk/testing | 1.2.2-snapshot.vc45c61c |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.0.3-snapshot.vc45c61c --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.4.2-snapshot.vc45c61c --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.10.6-snapshot.vc45c61c --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.0.0-snapshot.vc45c61c --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.0.20-snapshot.vc45c61c --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.22-snapshot.vc45c61c --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.2.7-snapshot.vc45c61c --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.6-snapshot.vc45c61c --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.19-snapshot.vc45c61c --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.1.10-snapshot.vc45c61c --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.2.2-snapshot.vc45c61c --save-exact
```

**Comment by octoper** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @octoper - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.0.3-snapshot.v999ed09 |
| @clerk/backend | 1.4.2-snapshot.v999ed09 |
| @clerk/elements | 0.10.6-snapshot.v999ed09 |
| @clerk/clerk-expo | 2.0.0-snapshot.v999ed09 |
| @clerk/express | 0.0.20-snapshot.v999ed09 |
| @clerk/fastify | 1.0.22-snapshot.v999ed09 |
| gatsby-plugin-clerk | 5.0.0-beta.45 |
| @clerk/nextjs | 5.2.7-snapshot.v999ed09 |
| @clerk/remix | 4.2.6-snapshot.v999ed09 |
| @clerk/clerk-sdk-node | 5.0.19-snapshot.v999ed09 |
| @clerk/tanstack-start | 0.1.10-snapshot.v999ed09 |
| @clerk/testing | 1.2.2-snapshot.v999ed09 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.0.3-snapshot.v999ed09 --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.4.2-snapshot.v999ed09 --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.10.6-snapshot.v999ed09 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.0.0-snapshot.v999ed09 --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@0.0.20-snapshot.v999ed09 --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@1.0.22-snapshot.v999ed09 --save-exact
```
`gatsby-plugin-clerk`
```sh
npm i gatsby-plugin-clerk@5.0.0-beta.45 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.2.7-snapshot.v999ed09 --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.6-snapshot.v999ed09 --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.19-snapshot.v999ed09 --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.1.10-snapshot.v999ed09 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.2.2-snapshot.v999ed09 --save-exact
```

**Review by panteliselef — APPROVED** [review]
